### ISSUE TITLE
Student Activity Dashboard

### ISSUE THREAD

**Issue by bmcutler** [Submitty/Submitty#5382]
**What problem are you trying to solve with Submitty**
Instructors may be required to submit periodic reports about student attendance/activity in online-only courses.  This information can be used to reach out for early intervention to students who are falling behind in their courses or struggling with time management.

**Describe the way you'd like to solve this problem**
Let's make a new "Activity Dashboard" page from the sidebar that is accessible only by instructor users.

Basic requirements:

Activity dashboard displays a table, 1 row per student
students are sorted by registration section, then alphabetically by last name, first name

Columns of with data for each student:
* last site login/logout (from access logs)
* last gradeable page load/access (from access logs)
* last gradeable submission (from access logs)
* last discussion forum access (from forum database)
* last discussion forum post (from forum database)

note: the column data value might be NULL
note: the login/logout data will be NULL if the student stays logged in through their browser (cookies don't expire)
note: performance concern -- some of these last access queries could be very expensive if the last access is many days ago -- have to look through many files

Also important:  Option to download this data to a .csv file.

Possible extra features:

* Add a date range at the top of the page.  start date: 1 week (7 days) ago. end date: defaults to the current time.  This significantly reduces the performance concern.

* Allow sorting by each column.

* Color coding to flag students who have been inactive.  Could just be no activity during the period, or no activity in a specific "important" column.

* Option to ask about page access/submission for a specific assignment (s)

* Option to ask about forum view/posts to a specific thread(s)

@MasterOdin -- any comments?

**Comment by bmcutler** [comment]
When #5383 is complete, the status of viewed course materials should also be included in the activity dashboard.

**Comment by MasterOdin** [comment]
We should not be scanning through the full log file for this kind of information. These files have a possibility of being rotated / pruned / whatever by a sysadmin and so they should not be taken as a potential "full source of truth", not to mention scanning through it is going to be slow.

All of this information should be pulled from the DB, though it will require creation of new tables in the course DB to capture the access information.

**Comment by bmcutler** [comment]
Here's a db query that gets the currently available data (version 1.0):

```
WITH
A AS
(SELECT registration_section, user_id, user_firstname, user_lastname
FROM users
ORDER BY registration_section, user_lastname, user_firstname, user_id),
B AS
(SELECT distinct on (user_id) user_id, timestamp
FROM gradeable_access where user_id is not NULL
ORDER BY user_id, timestamp desc),
C AS
(SELECT distinct on (user_id) user_id,submission_time
FROM electronic_gradeable_data
ORDER BY user_id, submission_time),
D AS
(SELECT distinct on (user_id) user_id, timestamp
FROM viewed_responses
ORDER BY user_id, timestamp desc),
E AS
(SELECT distinct on (author_user_id) author_user_id, timestamp
FROM posts
ORDER BY author_user_id, timestamp desc),
F AS
(SELECT student_id, count (student_id)
FROM poll_responses
GROUP BY student_id),
G AS
(SELECT distinct on (user_id) user_id, time_in
FROM queue
ORDER BY user_id, time_in desc)
SELECT
A.registration_section, A.user_id, user_firstname, user_lastname,
B.timestamp as gradeable_access,
C.submission_time as gradeable_submission,
D.timestamp as forum_view,
E.timestamp as forum_post,
F.count as num_poll_responses,
G.time_in as office_hours_queue
FROM
A
left join B on A.user_id=B.user_id
left join C on A.user_id=C.user_id
left join D on A.user_id=D.user_id
left join E on A.user_id=E.author_user_id
left join F on A.user_id=F.student_id
left join G on A.user_id=G.user_id
ORDER BY A.registration_section, A.user_lastname, A.user_firstname, A.user_id;
```

This dashboard page should show this type of data (formatted pretty -- using warning color to highlight the students with no "recent" activity), but also have an option to download the .csv (useful in huge classes).
